File: apps/android/app/src/main/java/com/kungsbackacarcommunity/app/live/FirebaseLiveLocationRepository.kt
package com.kungsbackacarcommunity.app.live

import android.content.Context
import com.google.firebase.FirebaseApp
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.google.firebase.functions.FirebaseFunctions
import com.kungsbackacarcommunity.app.garage.VehicleValidation
import java.time.Instant
import java.time.Year
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.suspendCancellableCoroutine

/**
 * [LiveLocationRepository] backed by the live.* callables (europe-west1) and a
 * Realtime Database value listener on the caller's own session node (Phase 12
 * slice 5).
 *
 * Callable names are the grouped-export forms live-startSession /
 * live-updatePosition / live-stopSession / live-hideMeNow
 * (contracts/functions/functions.json). Tasks are bridged to coroutines with
 * the same isActive-guarded pattern as the other repositories — no
 * play-services-await dependency. Construction is guarded ([createIfAvailable]
 * returns null when Firebase is not configured).
 */
class FirebaseLiveLocationRepository private constructor(
    private val functions: FirebaseFunctions,
    private val database: FirebaseDatabase,
) : LiveLocationRepository {

    override suspend fun startSession(duration: LiveSessionDuration) {
        call(START_SESSION, mapOf("duration" to duration.key))
    }

    override suspend fun updatePosition(coordinate: LiveCoordinate) {
        val coord =
            buildMap<String, Any> {
                put("latitude", coordinate.latitude)
                put("longitude", coordinate.longitude)
                put("recordedAt", coordinate.recordedAtIso)
                coordinate.accuracyMeters?.let { put("accuracyMeters", it) }
                coordinate.headingDegrees?.let { put("headingDegrees", it) }
                coordinate.speedMetersPerSecond?.let { put("speedMetersPerSecond", it) }
            }
        call(UPDATE_POSITION, mapOf("coordinate" to coord))
    }

    override suspend fun stopSession() {
        call(STOP_SESSION, mapOf("reason" to "user_stop"))
    }

    override suspend fun hideMeNow() {
        call(HIDE_ME_NOW, emptyMap())
    }

    override fun observeOwnSession(uid: String): Flow<LiveSessionInfo?> = callbackFlow {
        val ref = database.getReference("liveLocation/$uid/session")
        val listener =
            object : ValueEventListener {
                override fun onDataChange(snapshot: DataSnapshot) {
                    trySend(snapshot.toSessionInfo())
                }

                override fun onCancelled(error: DatabaseError) {
                    // Read denied/interrupted: surface "no session" instead of
                    // hanging; a later successful read self-corrects.
                    trySend(null)
                }
            }
        ref.addValueEventListener(listener)
        awaitClose { ref.removeEventListener(listener) }
    }

    override fun observeLatest(uid: String): Flow<LiveMarker?> = callbackFlow {
        // Per-uid marker read only (liveLocation/{uid}/latest); never scans the
        // collection. Mirrors observeOwnSession's callbackFlow/ValueEventListener.
        val ref = database.getReference("liveLocation/$uid/latest")
        val listener =
            object : ValueEventListener {
                override fun onDataChange(snapshot: DataSnapshot) {
                    trySend(snapshot.toLiveMarker(uid))
                }

                override fun onCancelled(error: DatabaseError) {
                    // Read denied (not an active, non-suspended member) or
                    // interrupted: emit null (no marker) rather than hanging; a
                    // later successful read self-corrects.
                    trySend(null)
                }
            }
        ref.addValueEventListener(listener)
        awaitClose { ref.removeEventListener(listener) }
    }

    private suspend fun call(name: String, data: Map<String, Any>): Unit =
        suspendCancellableCoroutine { continuation ->
            functions
                .getHttpsCallable(name)
                .call(data)
                .addOnCompleteListener { task ->
                    if (!continuation.isActive) return@addOnCompleteListener
                    if (task.isSuccessful) {
                        continuation.resume(Unit)
                    } else {
                        continuation.resumeWithException(
                            task.exception
                                ?: IllegalStateException("$name failed without a cause"),
                        )
                    }
                }
        }

    companion object {
        private const val REGION = "europe-west1"
        private const val START_SESSION = "live-startSession"
        private const val UPDATE_POSITION = "live-updatePosition"
        private const val STOP_SESSION = "live-stopSession"
        private const val HIDE_ME_NOW = "live-hideMeNow"

        fun createIfAvailable(context: Context): LiveLocationRepository? {
            if (FirebaseApp.getApps(context).isEmpty()) return null
            return FirebaseLiveLocationRepository(
                FirebaseFunctions.getInstance(REGION),
                FirebaseDatabase.getInstance(),
            )
        }
    }
}

/**
 * Maps the RTDB `latest` node to the Firebase-free [LiveMarker], or null when
 * absent (not sharing) or missing a coordinate. Reads only the marker-complete
 * fields written by live.updatePosition (latitude/longitude/displayName).
 */
private fun DataSnapshot.toLiveMarker(uid: String): LiveMarker? {
    if (!exists()) return null
    val latitude = child("latitude").getValue(Double::class.java) ?: return null
    val longitude = child("longitude").getValue(Double::class.java) ?: return null
    val displayName = child("displayName").getValue(String::class.java)
    return LiveMarker(
        uid = uid,
        latitude = latitude,
        longitude = longitude,
        displayName = displayName,
        mainCar = child("mainCar").toLiveMainCar(),
    )
}

/**
 * Maps the RTDB `mainCar` child (written by live.startSession's denormalization)
 * to [LiveMainCar], or null when absent or missing a required display field.
 * modelYear is read as a [Long] (RTDB stores integers as Long) and narrowed.
 */
private fun DataSnapshot.toLiveMainCar(): LiveMainCar? {
    if (!exists()) return null
    val make = child("make").getValue(String::class.java) ?: return null
    val model = child("model").getValue(String::class.java) ?: return null
    // Guard the Long before narrowing: a bare toInt() silently overflows on
    // malformed data (a client writing a huge/negative value would surface a
    // wrong year). Only accept plausible model years (same bounds the garage
    // form enforces); anything outside drops the whole main car rather than
    // render a bogus year.
    val modelYear =
        child("modelYear").getValue(Long::class.java)?.let { raw ->
            val minYear = VehicleValidation.MIN_MODEL_YEAR.toLong()
            val maxYear = VehicleValidation.maxModelYear(Year.now().value).toLong()
            if (raw in minYear..maxYear) raw.toInt() else null
        } ?: return null
    val imagePath = child("imagePath").getValue(String::class.java)
    return LiveMainCar(make = make, model = model, modelYear = modelYear, imagePath = imagePath)
}

/** Maps the RTDB session node to the Firebase-free [LiveSessionInfo]. */
private fun DataSnapshot.toSessionInfo(): LiveSessionInfo? {
    if (!exists()) return null
    val status = LiveSessionStatus.fromWire(child("status").getValue(String::class.java)) ?: return null
    val sessionId = child("id").getValue(String::class.java) ?: return null
    val duration = LiveSessionDuration.fromKey(child("duration").getValue(String::class.java))
    val expiresAtMillis =
        child("expiresAt").getValue(String::class.java)?.let { iso ->
            try {
                Instant.parse(iso).toEpochMilli()
            } catch (parse: Exception) {
                null
            }
        }
    return LiveSessionInfo(sessionId, status, duration, expiresAtMillis)
}

File: apps/ios/KCC/Live/FirebaseLiveLocationRepository.swift
import FirebaseAuth
import FirebaseCore
import FirebaseDatabase
import FirebaseStorage
import Foundation

/// ``LiveLocationRepository`` backed by the `live.*` callables and a Realtime
/// Database listener on the caller's own session node — the iOS port of
/// Android's `live/FirebaseLiveLocationRepository.kt` (own-session slice).
///
/// The split is Android's exactly: WRITES go through the callables
/// (`live-startSession` / `live-updatePosition` / `live-stopSession` /
/// `live-hideMeNow`, all in europe-west1 via ``KccFunctionsClient``); the
/// RTDB `liveLocation/{uid}` nodes are backend-written, and this class only
/// READS the owner's session node (owner-only read —
/// firebase/database.rules.json).
///
/// Callable failures surface as ``KccFunctionsError`` (contract code only —
/// never the SDK message, which may reference the request payload; per the
/// privacy rules a payload here contains exact GPS coordinates, so nothing
/// from a failure is ever logged). Construction is guarded
/// (``createIfAvailable()`` returns nil without Firebase config), mirroring
/// the other repositories and Android's `createIfAvailable`.
final class FirebaseLiveLocationRepository: LiveLocationRepository, @unchecked Sendable {
    private let functions: KccFunctionsClient
    private let database: Database
    private let storage: Storage

    private init(functions: KccFunctionsClient, database: Database, storage: Storage) {
        self.functions = functions
        self.database = database
        self.storage = storage
    }

    // MARK: - Callables (writes)

    func startSession(duration: LiveSessionDuration, vehicleId: String?) async throws {
        var payload: [String: Any] = ["duration": duration.key]
        // Only send the car when one was chosen; omitting it lets the server
        // fall back to the main car (the callable schema treats a blank
        // string as invalid, so never send an empty one) — Android parity.
        if let vehicleId, !vehicleId.isEmpty {
            payload["vehicleId"] = vehicleId
        }
        _ = try await functions.call(Self.startSessionCallable, payload: payload)
    }

    func updatePosition(_ coordinate: LiveCoordinate) async throws {
        var coord: [String: Any] = [
            "latitude": coordinate.latitude,
            "longitude": coordinate.longitude,
            "recordedAt": coordinate.recordedAtIso,
        ]
        if let accuracy = coordinate.accuracyMeters { coord["accuracyMeters"] = accuracy }
        if let heading = coordinate.headingDegrees { coord["headingDegrees"] = heading }
        if let speed = coordinate.speedMetersPerSecond { coord["speedMetersPerSecond"] = speed }
        _ = try await functions.call(Self.updatePositionCallable, payload: ["coordinate": coord])
    }

    func stopSession() async throws {
        _ = try await functions.call(Self.stopSessionCallable, payload: ["reason": "user_stop"])
    }

    func hideMeNow() async throws {
        _ = try await functions.call(Self.hideMeNowCallable, payload: [:])
    }

    // MARK: - RTDB (own-session read)

    func ownSessionUpdates(uid: String) -> AsyncStream<LiveSessionInfo?> {
        let ref = database.reference(withPath: "liveLocation/\(uid)/session")
        return AsyncStream(bufferingPolicy: .bufferingNewest(1)) { continuation in
            let handle = ref.observe(
                .value,
                with: { snapshot in
                    let map = snapshot.value as? [String: Any]
                    continuation.yield(map.flatMap(LiveSessionInfo.fromMap))
                },
                withCancel: { _ in
                    // Read denied/interrupted: surface "no session" instead
                    // of hanging; a later successful read self-corrects
                    // (Android's onCancelled posture). The error is never
                    // logged — it can embed the database path with the uid.
                    continuation.yield(nil)
                }
            )
            let box = ObserverBox(reference: ref)
            box.setHandle(handle)
            continuation.onTermination = { _ in
                box.removeObserverOnce()
            }
        }
    }

    func latestUpdates(uid: String) -> AsyncStream<LiveMarker?> {
        let events = latestUpdateEvents(uid: uid)
        return AsyncStream(bufferingPolicy: .bufferingNewest(1)) { continuation in
            let task = Task {
                for await event in events {
                    if Task.isCancelled { break }
                    guard case let .value(marker) = event else { continue }
                    continuation.yield(marker)
                }
                continuation.finish()
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    func latestUpdateEvents(uid: String) -> AsyncStream<LiveMarkerUpdateEvent> {
        let ref = database.reference(withPath: "liveLocation/\(uid)/latest")
        return AsyncStream(bufferingPolicy: .bufferingNewest(1)) { continuation in
            var handle: DatabaseHandle?
            let box = ObserverBox(reference: ref)
            handle = ref.observe(
                .value,
                with: { snapshot in
                    let map = snapshot.value as? [String: Any]
                    continuation.yield(.value(map.flatMap { LiveMarker.fromMap(uid: uid, $0) }))
                },
                withCancel: { error in
                    box.removeObserverOnce()
                    if LiveRealtimeRetryPolicy.shouldRetry(error) {
                        continuation.yield(.retry)
                    } else {
                        continuation.yield(.value(nil))
                    }
                    continuation.finish()
                }
            )
            box.setHandle(handle)
            continuation.onTermination = { _ in
                box.removeObserverOnce()
            }
        }
    }

    func imageDownloadURL(for imagePath: String) async -> URL? {
        try? await storage.reference(withPath: imagePath).downloadURL()
    }

    func currentUserId() -> String? {
        Auth.auth().currentUser?.uid
    }

    // MARK: - Factory

    // Grouped-export callable names — suffixed so call sites read as
    // "invoke the named callable", not as recursion into the instance
    // methods they back.
    private static let startSessionCallable = "live-startSession"
    private static let updatePositionCallable = "live-updatePosition"
    private static let stopSessionCallable = "live-stopSession"
    private static let hideMeNowCallable = "live-hideMeNow"

    private static let cachedLock = NSLock()
    nonisolated(unsafe) private static var cached: FirebaseLiveLocationRepository?

    /// Returns the process-wide repository when Firebase is configured for
    /// this build, or nil when GoogleService-Info.plist is absent (CI, local
    /// validation builds — see apps/ios/README.md).
    ///
    /// Emulator seams follow the shared `FIREBASE_*_EMULATOR_HOST`
    /// convention: `FIREBASE_DATABASE_EMULATOR_HOST` (port 9000 per
    /// firebase.json) points the Realtime Database SDK at the emulator
    /// before first use; the callables ride ``KccFunctionsClient``'s own
    /// `FIREBASE_FUNCTIONS_EMULATOR_HOST` seam, while
    /// `FIREBASE_STORAGE_EMULATOR_HOST` resolves convoy car photos.
    static func createIfAvailable() -> LiveLocationRepository? {
        guard FirebaseApp.app() != nil else { return nil }
        cachedLock.lock()
        defer { cachedLock.unlock() }
        if let cached { return cached }
        guard let functions = KccFunctionsClient.createIfAvailable() else { return nil }
        let database: Database
        if let emulator = FirebaseEmulatorHost.parse(
            ProcessInfo.processInfo.environment["FIREBASE_DATABASE_EMULATOR_HOST"]
        ) {
            // Point at the emulator's namespace-less local instance; must
            // happen before the instance hands out references.
            database = Database.database()
            database.useEmulator(withHost: emulator.host, port: emulator.port)
        } else {
            database = Database.database()
        }
        let storage = Storage.storage()
        if let emulator = FirebaseEmulatorHost.parse(
            ProcessInfo.processInfo.environment["FIREBASE_STORAGE_EMULATOR_HOST"]
        ) {
            storage.useEmulator(withHost: emulator.host, port: emulator.port)
        }
        let repository = FirebaseLiveLocationRepository(
            functions: functions,
            database: database,
            storage: storage
        )
        cached = repository
        return repository
    }
}

/// `DatabaseReference`/handle pair carried into the stream's `onTermination`
/// closure, which must be Sendable — all it does is remove the observer,
/// which the Database SDK documents as thread-safe (same pattern as the
/// Firestore `ListenerBox`es).
private final class ObserverBox: @unchecked Sendable {
    let reference: DatabaseReference
    private let lock = NSLock()
    private var handle: DatabaseHandle?

    init(reference: DatabaseReference) {
        self.reference = reference
    }

    func setHandle(_ handle: DatabaseHandle?) {
        lock.lock()
        self.handle = handle
        lock.unlock()
    }

    func removeObserverOnce() {
        lock.lock()
        let handle = self.handle
        self.handle = nil
        lock.unlock()
        if let handle {
            reference.removeObserver(withHandle: handle)
        }
    }
}

private enum LiveRealtimeRetryPolicy {
    private static let operationFailedCode = -2
    private static let disconnectedCode = -4
    private static let networkErrorCode = -24
    private static let accessFragments = ["permission", "denied", "unauthor", "access"]

    static func shouldRetry(_ error: Error) -> Bool {
        let nsError = error as NSError
        let details = [
            nsError.localizedDescription,
            nsError.localizedFailureReason,
            nsError.userInfo[NSDebugDescriptionErrorKey] as? String
        ]
        .compactMap { $0?.lowercased() }
        if details.contains(where: { detail in
            accessFragments.contains { detail.contains($0) }
        }) {
            return false
        }
        return nsError.code == disconnectedCode
            || nsError.code == networkErrorCode
            || nsError.code == operationFailedCode
    }
}
